### PR TITLE
Greatly improve performance by caching subsequent glob calls

### LINKED ISSUE #34024: [Issue] Greatly improve performance by caching subsequent glob calls
This issue is automatically created based on existing pull request: magento/magento2#33998: Greatly improve performance by caching subsequent glob calls

---------

### Description (*)
This pull request introduces an in-memory cache for glob system calls. On systems where file I/O is slow (for example Docker for Mac), glob system calls can be quite slow. This pull requests caches the same glob calls and returns a cached result if possible. In my testing this reduced the glob calls on the category listings from ~8K to ~1.6K.

Blackfire comparison:

Screenshot 2021-09-06 at 09 51 21

### Related Pull Requests
None

### Fixed Issues (if relevant)
None
1. Fixes magento/magento2#<issue_number>

### Manual testing scenarios (*)
1. Profile any page before this diff is applied
2. Apply diff and reprofile
3. See a reduction in `glob()` calls

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [x] All automated tests passed successfully (all builds are green)

### LINKED ISSUE #34025: [Issue] Greatly improve performance by caching subsequent glob calls
This issue is automatically created based on existing pull request: magento/magento2#33998: Greatly improve performance by caching subsequent glob calls

---------

### Description (*)
This pull request introduces an in-memory cache for glob system calls. On systems where file I/O is slow (for example Docker for Mac), glob system calls can be quite slow. This pull requests caches the same glob calls and returns a cached result if possible. In my testing this reduced the glob calls on the category listings from ~8K to ~1.6K.

Blackfire comparison:

Screenshot 2021-09-06 at 09 51 21

### Related Pull Requests
None

### Fixed Issues (if relevant)
None
1. Fixes magento/magento2#<issue_number>

### Manual testing scenarios (*)
1. Profile any page before this diff is applied
2. Apply diff and reprofile
3. See a reduction in `glob()` calls

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [x] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#34024: Greatly improve performance by caching subsequent glob calls

### PR DESCRIPTION
**Pull request by thlassche**
### Description (*)
This pull request introduces an in-memory cache for glob system calls. On systems where file I/O is slow (for example Docker for Mac), glob system calls can be quite slow. This pull requests caches the same glob calls and returns a cached result if possible. In my testing this reduced the glob calls on the category listings from ~8K to ~1.6K.

Blackfire comparison:

Screenshot 2021-09-06 at 09 51 21

### Related Pull Requests
None

### Fixed Issues (if relevant)
None
1. Fixes magento/magento2#<issue_number>

### Manual testing scenarios (*)
1. Profile any page before this diff is applied
2. Apply diff and reprofile
3. See a reduction in `glob()` calls

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [x] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#34024: Greatly improve performance by caching subsequent glob calls

### Resolved issues:
1. [x] resolves magento/magento2#34025: Greatly improve performance by caching subsequent glob calls

### CODE REVIEW

**Comment by ihor-sviziev** [comment]
@magento run all tests

File: lib/internal/Magento/Framework/Filesystem/Glob.php [hunk 1]
```diff
@@ -14,6 +14,7 @@
  */
 class Glob extends LaminasGlob
 {
+    protected static $cache = [];
```
Comment (ihor-sviziev) [inline lib/internal/Magento/Framework/Filesystem/Glob.php]: It looks like it should be private cache

```suggestion
    private static $cache = [];
```

Also, wonder if we should have a mechanism for removing the cache...
Will see after test results

**Comment by thlassche** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
It might significantly improve performance not only on dockerized environments but also on the ec2 instances or any server using network file storage as magento storage.

**Review by andrewbess — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @andrewbess, thank you for the review.
[ENGCOM-9214](https://jira.corp.magento.com/browse/ENGCOM-9214) has been created to process this Pull Request
:eight_spoked_asterisk: @andrewbess, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-9214](https://jira.corp.magento.com/browse/ENGCOM-9214) has been created to process this Pull Request
:eight_spoked_asterisk: @ihor-sviziev, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by ihor-sviziev** [comment]
I think we don't need to cover it with any test as it will be not so easy.

**Review by Den4ik — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @Den4ik, thank you for the review.
[ENGCOM-9214](https://jira.corp.magento.com/browse/ENGCOM-9214) has been created to process this Pull Request

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento create issue

**Comment by engcom-Hotel** [comment]
@magento create issue

**Comment by engcom-Hotel** [comment]
@magento give me 2.4-develop instance

**Comment by engcom-Hotel** [comment]
@magento give me 2.4-develop instance

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE

**Comment by onlinebizsoft** [comment]
> It might significantly improve performance not only on dockerized environments but also on the ec2 instances or any server using network file storage as magento storage.

What do you mean by "magento storage"? You mean "media storage"? @ihor-sviziev

**Comment by ihor-sviziev** [comment]
@onlinebizsoft, I think it's not only media folder, but also where magento code is located.

**Review by engcom-Hotel — CHANGES_REQUESTED** [review]
Hello @thlassche,

Some tests are failing for e.g. Functional Tests B2B, Functional Tests CE, Functional Tests EE. Please look into this.

**Comment by ihor-sviziev** [comment]
@engcom-Hotel, the service that responsible for the currency API doesn't work. Here is confirmation: https://github.com/magento/magento2/pull/34021#issuecomment-915258946

**Comment by onlinebizsoft** [comment]
> @onlinebizsoft, I think it's not only media folder, but also where magento code is located.

Interesting! Im not sure how glob is used and called in Magento. I thought Magento files would be loaded in OPcache memory already

**Comment by engcom-Hotel** [comment]
> @engcom-Hotel, the service that responsible for the currency API doesn't work. Here is confirmation: [#34021 (comment)](https://github.com/magento/magento2/pull/34021#issuecomment-915258946)

Thanks for the clarification @ihor-sviziev

**Comment by engcom-Hotel** [comment]
✔️ QA Passed

Hello @thlassche,

Thanks for the contribution!

**Preconditions**
Magento 2.4-develop branch

**Manual testing scenario**
1. Run profiler on Magento 2.4-develop with sample data installed.
2. Run profiler on this PR fix with sample data installed. 
3. Compare results.

image

**Comment by onlinebizsoft** [comment]
> ✔️ QA Passed
> 
> Hello @thlassche,
> 
> Thanks for the contribution!
> 
> **Preconditions**
> Magento 2.4-develop branch
> 
> **Manual testing scenario**
> 
>     1. Run profiler on Magento 2.4-develop with sample data installed.
> 
>     2. Run profiler on this PR fix with sample data installed.
> 
>     3. Compare results.
> 
> 
> image

Just reduce few ms?!!

**Comment by thlassche** [comment]
> > ✔️ QA Passed
> 
> > 
> 
> > Hello @thlassche,
> 
> > 
> 
> > Thanks for the contribution!
> 
> > 
> 
> > **Preconditions**
> 
> > Magento 2.4-develop branch
> 
> > 
> 
> > **Manual testing scenario**
> 
> > 
> 
> >     1. Run profiler on Magento 2.4-develop with sample data installed.
> 
> > 
> 
> >     2. Run profiler on this PR fix with sample data installed.
> 
> > 
> 
> >     3. Compare results.
> 
> > 
> 
> > 
> 
> > image
> 
> 
> 
> Just reduce few ms?!!

It depends a lot on what is currently in the cache and what is not. The first load of uncached pages will benefit way more.

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE

**Comment by Den4ik** [comment]
@magento run all test

**Comment by Den4ik** [comment]
@magento run all tests

**Pull request merged**